implement non-user-specific caching functionality

cellog opened the pull request:
# Description

In order to cache locks, which are not specific to any user, but are specific to networks, this PR introduces `getReadOnly` and `putReadOnly`.

In order to cache account/network, both of which are used as keys for everything else, this PR also introduces specific caching methods `getAccount` `getNetwork` and their setters.

# Issues

<!-- This PR should fix or reference at least one existing issue ID. Add or delete as appropriate. -->

Refs #3138 

# Checklist:

- [X] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [X] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [X] My code follows the style guidelines of this project, including naming conventions
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [X] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

<!--
PS: [Read how to write the perfect pull request](https://blog.github.com/2015-01-21-how-to-write-the-perfect-pull-request/)
-->

cellog commented on paywall/src/data-iframe/cache/localStorage.js:
```diff
@@ -43,6 +43,40 @@ function getContainer(window, key, type = false) {
+  return get(
+    window,
+    networkId,
+    '0x0000000000000000000000000000000000000000',
```
I'm going to add a comment in the code about why I chose address 0 (nobody can use this, so it's a guaranteed readonly value)

akeem commented on paywall/src/__tests__/data-iframe/cache/localStorage-no-localStorage.test.js:
```diff
@@ -14,6 +24,16 @@ describe('localStorage cache', () => {
+      try {
+        await getReadOnly()
+      } catch (e) {
+        expect(e.message).toBe('Cannot get value from localStorage')
```
Depending on the types of errors possible it might be worth while to move the  error message in the test to its own variable to be referenced in the subsequent tests

akeem commented on paywall/src/data-iframe/cache/localStorage.js:
```diff
@@ -43,6 +45,42 @@ function getContainer(window, key, type = false) {
+ * @param {int} networkId the ethereum network id
+ * @param {string} type the type of account data to retrieve
+ */
+export async function getReadOnly(window, networkId, type) {
```
Is there some context around these methods ending with `ReadOnly`?

cellog replied:
caching read-write values (keys, transactions) depends on a user account. read-only values are currently locks, which are account-independent

akeem commented on paywall/src/data-iframe/cache/localStorage.js:
```diff
@@ -86,6 +124,58 @@ export async function put(window, networkId, accountAddress, type, value) {
+ * @returns {string}
+ */
+export async function getAccount(window) {
+  if (!localStorageAvailable(window)) {
```
It might be worthwhile to extract this into a method and us it in the methods calling the functionality below

akeem commented on paywall/src/data-iframe/cache/localStorage.js:
```diff
@@ -86,6 +124,58 @@ export async function put(window, networkId, accountAddress, type, value) {
+ * @param {object} window this is the global context, either global, window, or self
+ * @param {number} network the id of the current ethereum network
+ */
+export async function setNetwork(window, network) {
```
Not blocking but I think that we could probably generalize these getters and setters

akeem commented on paywall/src/__tests__/data-iframe/cache/localStorage.test.js:
```diff
@@ -89,6 +97,58 @@ describe('localStorage cache', () => {
+      expect(await getReadOnly(fakeWindow, 123, 'there')).toEqual('hello')
+    })
+
+    it('value is set, but malformed', async () => {
```
Is the setting of malformed data handled at a layer below the code in this pull request?

cellog replied:
localStorage is publicly accessible, so this guards against a malicious script trying to muck things up

cellog commented:
ok @akeem I refactored to add `ensureLocalStorageAvailable` for generic error-handling with more specific error messages too (double win). I also removed the `ReadOnly` stuff, and instead changed the parameters of `get` and `put` (and others) to take a destructured object. This is because I already had `type` as an optional parameter, and this is the simplest way to have 2 optional parameters.

PTAL when you get a chance.
